core/dbt/context/runtime.py

```diff
@@ -8,80 +8,79 @@
 from dbt.logger import GLOBAL_LOGGER as logger  # noqa
 
 
-execute = True
-
-
-def ref(db_wrapper, model, config, manifest):
-    current_project = config.project_name
-    adapter = db_wrapper.adapter
-
-    def do_ref(*args):
-        target_model_name = None
-        target_model_package = None
+class BaseRefResolver(dbt.context.common.BaseResolver):
+    def resolve(self, args):
+        name = None
+        package = None
 
         if len(args) == 1:
-            target_model_name = args[0]
+            name = args[0]
         elif len(args) == 2:
-            target_model_package, target_model_name = args
+            package, name = args
         else:
-            dbt.exceptions.ref_invalid_args(model, args)
+            dbt.exceptions.ref_invalid_args(self.model, args)
 
         target_model = ParserUtils.resolve_ref(
-            manifest,
-            target_model_name,
-            target_model_package,
-            current_project,
-            model.get('package_name'))
+            self.manifest,
+            name,
+            package,
+            self.current_project,
+            self.model.package_name)
 
         if target_model is None or target_model is ParserUtils.DISABLED:
             dbt.exceptions.ref_target_not_found(
-                model,
-                target_model_name,
-                target_model_package)
-
-        target_model_id = target_model.get('unique_id')
-
-        if target_model_id not in model.get('depends_on', {}).get('nodes'):
-            dbt.exceptions.ref_bad_context(model,
-                                           target_model_name,
-                                           target_model_package)
-
-        is_ephemeral = (get_materialization(target_model) == 'ephemeral')
-
-        if is_ephemeral:
-            model.set_cte(target_model_id, None)
-            return adapter.Relation.create(
-                type=adapter.Relation.CTE,
-                identifier=add_ephemeral_model_prefix(
-                    target_model_name)).quote(identifier=False)
+                self.model,
+                name,
+                package)
+        return target_model, name
+
+    def create_ephemeral_relation(self, target_model, name):
+        self.model.set_cte(target_model.unique_id, None)
```

Comment:
i think this line fails if `self.model` is a `ParsedMacro`. That only happens if you try to `ref` an ephemeral model in an operation AFAIK. I'm happy to not support that (can't think of a reason anyone would want it, or what it would compile to) but it might be possible to improve the error message here:

```
Encountered an uncaught exception while running operation: 'ParsedMacro' object has no attribute 'set_cte'
```

Reply:
Interesting find!

I think we should fix this in a future release, it's part of the same root problem as this was: `ParsedMacro`s are not _quite_ the correct expression for this problem in dbt, even if it is just running a macro. If nothing else, explaining this limitation all the time will surely get tedious.

I've always felt what really should be going on here is something like, we get the parameters and do something similar to what we do in tests, and given `dbt run-operation my_macro --args '{...}'`, emit:

```
{% magic %}
{{ my_macro(**args) }}
{% endmagic %}
```

(you can imagine `s/magic/operation/g`, but I wanted to avoid that word)

What gets run through `run-operation` is in that world is something that's a bit more than a `ParsedMacro`, which seems totally correct, and a lot like the RPC behavior (which makes it even more correct, right?).

On the other hand, it also seems more involved than I really want to get into at the tail end of a release, especially one that's already as big as this one!

Reply:
For this PR, maybe I can add a `ParsedMacro.set_cte` method that raises an exception saying you can't do ephemeral refs in run-operation? I'd feel pretty good about that.

Reply:
I did the above, except I just overrode `operation.RefResolver.create_ephemeral_relation` to do the raising, since that's the only place that it matters.